### PR TITLE
Issue #1777: Allow blanking of first and/or second nameserver if enough entries are present (rjm)

### LINKED ISSUE #1777: Clearing out Nameserver record should be the same as using the delete icon
### Issue description

Clearing out Nameserver record should be the same as using the delete icon. Currently, it is not the same. Clearing nameserver record in row 1 and row 2 and saving results in the “at least two nameservers required” message. But deleting the rows using the trash icon and saving doesn’t result in the message.

See [slack thread](https://cisa-corp.slack.com/archives/C05BGB4L5NF/p1707918276148259) for steps to reproduce issue and troubleshooting discussion.

image

### Acceptance criteria

- [ ] Clearing out Nameserver record should be the same as using the delete icon.

### Additional context

_No response_

### Links to other issues

Related: https://github.com/cisagov/manage.get.gov/issues/1192

### PR DESCRIPTION
**Pull request by rachidatecs**
## Ticket

Resolves #1777 

## Changes

- Don't let the nameserver form clean remove the entered data
- Use the data in the nameserver formset clean to see whether we have enough entries
- Fix an undiscovered bug where if you submit a formset with only 2 visible forms that subsequently errors, you cannot add a new form to the formset after the failed submission
- Fix an undiscovered bug where if you submit a formset with only 2 visible forms that subsequently errors, you can still click on 'delete' for forms 1 and 2 (that delete should be disabled and unclickable)

## Context for reviewers

Make sure to test for the 2 undiscovered bug fixes mentioned above.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [ ] Pulled this branch locally and tested it
- [ ] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [ ] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the associated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [ ] Interactions with external systems are wrapped in try/except
- [ ] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/assets/js/get-gov.js [hunk 1]
```diff
@@ -530,7 +530,7 @@ function hideDeletedForms() {
   let isDotgovDomain = document.querySelector(".dotgov-domain-form");
   // The Nameservers formset features 2 required and 11 optionals
   if (isNameserversForm) {
-    cloneIndex = 2;
+    // cloneIndex = 2;
```
Comment (rachidatecs) [inline src/registrar/assets/js/get-gov.js:533]: I think this was needed in an earlier iteration of the code. Right now it's in the way and keeps you from adding a form if you load the page with only 2 forms (like after a failed submission). Keeping it commented out until the dust clears on the PR

File: src/registrar/assets/js/get-gov.js [hunk 2]
```diff
@@ -766,3 +766,21 @@ function toggleTwoDomElements(ele1, ele2, index) {
...
+  let isNameserversForm = document.querySelector(".nameservers-form");
+  if (isNameserversForm) {
+    let forms = document.querySelectorAll(".repeatable-form");
+    if (forms.length < 3) {
+      // Hide the delete buttons on the 2 nameservers
+      forms.forEach((form) => {
+        Array.from(form.querySelectorAll('.delete-record')).forEach((deleteButton) => {
+          deleteButton.setAttribute("disabled", "true");
+        });
+      });
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/assets/js/get-gov.js:783]: I don't think it's necessary. The form validation is a gate that protects the data This stuff is UI niceties.

File: src/registrar/forms/domain.py [hunk 3]
```diff
@@ -149,6 +158,19 @@ def clean(self):
...
+            1 for form in self.forms if form.cleaned_data.get("server") and form.cleaned_data.get("server").strip()
+        )
+        if valid_servers_count >= 2:
+            # If there are, remove the "At least two name servers are required" error from each form
+            # This will allow for successful submissions when the first or second entries are blanked
+            # but there are enough entries total
+            for form in self.forms:
+                if form.errors.get("server") == ["At least two name servers are required."]:
+                    form.errors.pop("server")
+
```
[the remark this answers is not in the document]
Comment (rachidatecs) [reply src/registrar/forms/domain.py:173]: Yes. You need to be aware of all forms in the formset, so you have to do it in a formset method. Django does a first pass on each form first and we shouldn't mess with validation there, out of context.

**Review by kristinacyin — APPROVED** [review]
LGTM! 🎆

**Pull request merged**